src/core/ChannelHandler.js

```diff
@@ -0,0 +1,103 @@
+import Observable from 'zen-observable';
+
+const events = [
+  'phx_reply',
+  'phx_error',
+  'status_changed',
+  'stopped',
+  'error',
+  'failed',
+  'starting',
+  'started',
+  'deploying',
+  'deployed',
+  'deployment_failed',
+  'ready',
+  'terminated',
+  'snapshot_taken',
+  'snapshot_reverted'
+];
+
+export default class ChannelHandler {
+  constructor(name, socket, globalLogger) {
+    this._name = name;
+    this._channel = socket.channel(this._name);
+    this._stream = this._buildChannelStream(events);
+    this._log = globalLogger.extend(`${this._name}`);
+    this._logger = this._stream.subscribe(value =>
+      this._log(JSON.stringify(value, null, 4))
+    );
+    this.init();
+  }
+
+  _buildChannelStream(eventsList) {
+    return new Observable(stream => {
+      const setChannelEvent = event => {
+        this._channel.on(event, payload => {
+          switch (event) {
+            case 'deployment_failed':
+              stream.error({ event, payload });
+              break;
+            case 'error':
+              stream.error({ event, payload });
+              break;
+            case 'failed':
+              stream.error({ event, payload });
+              break;
+            case 'status_changed':
+              stream.next({
+                event: `status_changed_${payload.data}`,
+                payload
+              });
+              break;
+            default:
+              stream.next({ event, payload });
+          }
+        });
+      };
+
+      eventsList.forEach(event => {
+        setChannelEvent(event);
+      });
+    });
+  }
+
+  on(event, cb) {
+    const ref = this._channel.on(event, data => {
+      const off = () => this._channel.off(event, ref);
+      cb(data, off);
+    });
+  }
+
+  once(_event) {
+    return new Promise((resolve, reject) => {
+      const observer = this._stream.subscribe(
+        ({ event, payload }) => {
+          if (event === _event) {
+            observer.unsubscribe();
+            resolve({ event, payload });
+          }
+        },
+        ({ event, payload }) => {
+          reject(JSON.stringify({ event, payload }, null, 4));
+        }
+      );
+    });
+  }
+
+  init() {
+    this._channel.join();
+  }
+
+  push(event, payload = {}) {
```

Comment:
how/where is this being used?
Also, let's avoid using `event` as a variable in our code, it's not a reserved word per se but there are gotchas we shouldn't have to think about.

Reply:
Used in `Client.js` for referencing the `push()` function on the `_channel` object under each `channelHandler`

So when we call `client.create(options)`:

```javascript
client.channel('api').push(options)
...
client._socket._channels.api._channel.push(options)
```
So `client._socket._channels` is a mapping of all `ChannelHandlers`, one for each chain instance. Each `ChannelHandler` has a phoenix.js channel instantiated under `_channel`. We can call `push` on this from by referencing `_channel` directly but I thought it would be cleaner to be able `push()` directly from the `ChannelHandler`

Reply:
Will change event, would `action` be preferred? This could also be a set of constants which I think would be a cleaner solution? `ChainActions`

Reply:
ok...

IMO `eventName` is a fine argument name here.